File: oop_concepts/abstraction.py
from abc import ABC, abstractmethod

class UPI(ABC):
    @abstractmethod
    def pay(self, amount):
        pass

class PhonePe(UPI):
    def pay(self, amount):
        print(f"Processing payment of {amount} via PhonePe...")
        print("Payment successful!")

class GooglePay(UPI):
    def pay(self, amount):
        print(f"Processing payment of {amount} via GooglePay...")
        print("Payment successful!")

if __name__ == "__main__":
    upi_abstraction = PhonePe()
    upi_abstraction.pay(500)

    upi_abstraction = GooglePay()
    upi_abstraction.pay(1000)

File: oop_concepts/abstraction.cs
// Abstraction using interface
public interface IUPI
{
    void Pay(float amount);
}

// PhonePe implements UPI abstraction
public class PhonePe : IUPI
{
    public void Pay(float amount)
    {
        Console.WriteLine($"Processing payment of {amount} via PhonePe...");
        Console.WriteLine("Payment successful!");
    }
}

public class GooglePay : IUPI
{
    public void Pay(float amount)
    {
        Console.WriteLine($"Processing payment of {amount} via GooglePay...");
        Console.WriteLine("Payment successful!");
    }
}

public class Program
{
    public static void Main()
    {
        IUPI upiAbstraction;

        upiAbstraction = new PhonePe();
        upiAbstraction.Pay(500);

        upiAbstraction = new GooglePay();
        upiAbstraction.Pay(1000);
    }
}